clients/src/main/java/org/apache/kafka/clients/ClientTelemetryState.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.clients;
+
+import org.apache.kafka.common.errors.IllegalClientTelemetryStateException;
+import org.apache.kafka.common.utils.Utils;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * State that helps determine where client exists in the telemetry state i.e. subscribe->wait->push loop.
+ */
+public enum ClientTelemetryState {
+
+    /**
+     * Client needs subscription from the broker.
+     */
+    SUBSCRIPTION_NEEDED,
+
+    /**
+     * Network I/O is in progress to retrieve subscription.
+     */
+    SUBSCRIPTION_IN_PROGRESS,
+
+    /**
+     * Awaiting telemetry interval for pushing metrics to broker.
+     */
+    PUSH_NEEDED,
+
+    /**
+     * Network I/O in progress for pushing metrics payload.
+     */
+    PUSH_IN_PROGRESS,
+
+    /**
+     * Need to push the terminal metrics payload.
+     */
+    TERMINATING_PUSH_NEEDED,
+
+    /**
+     * Network I/O in progress for pushing terminal metrics payload.
+     */
+    TERMINATING_PUSH_IN_PROGRESS,
+
+    /**
+     * No more work should be performed, telemetry client terminated.
+     */
+    TERMINATED;
+
+    private final static Map<ClientTelemetryState, List<ClientTelemetryState>> VALID_NEXT_STATES = new EnumMap<>(ClientTelemetryState.class);
+
+    static {
+        /*
+         If clients needs a subscription, then issue telemetry API to fetch subscription from broker.
+
+         However, it's still possible that client doesn't get very far before terminating.
+        */
+        VALID_NEXT_STATES.put(
+            SUBSCRIPTION_NEEDED, Arrays.asList(SUBSCRIPTION_IN_PROGRESS, TERMINATING_PUSH_NEEDED,
```

Comment:
I don't see why TERMINATING_PUSH_NEEDED is a valid next state from SUBSCRIPTION_NEEDED.

Reply:
I think you are right as there is nothing to push. Removed it, thanks.